hw/ip/kmac/dv/dpi/digestpp_dpi.cpp

```diff
@@ -0,0 +1,276 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <cstdint>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "svdpi.h"
+#include "vendor/kerukuro_digestpp/algorithm/kmac.hpp"
+#include "vendor/kerukuro_digestpp/algorithm/sha3.hpp"
+#include "vendor/kerukuro_digestpp/algorithm/shake.hpp"
+
+extern "C" {
+
+// TODO(udi) might need to implement endian conversion
+
+//////////////////////
+// HELPER FUNCTIONS //
+//////////////////////
+
+/**
+ * Generic function to load an unsized array from SV memory into C memory.
+ */
+static void load_arr_from_simulator(const svOpenArrayHandle arr,
+                                    uint8_t *array_out,
+                                    uint64_t array_len) {
+  uint8_t *arr_ptr;
+
+  if (array_len == 0) {
+    return;
+  }
+
+  arr_ptr = (uint8_t *)svGetArrayPtr(arr);
+  memcpy(array_out, arr_ptr, array_len);
+}
+
+/**
+ * Generic function to write an unsized array from C memory into SV memory.
+ */
+static void write_array_to_simulator(const svOpenArrayHandle arr,
+                                     uint8_t *data) {
+  svBitVecVal data_val;
+  uint64_t arr_len = svSize(arr, 1);
+
+  for (uint64_t i = 0; i < arr_len; ++i) {
+    data_val = (svBitVecVal)data[i];
+    svPutBitArrElem1VecVal(arr, &data_val, i);
+  }
+}
+
+/**
+ * Helper function to calculate generic length SHA3 algorithm.
+ *
+ * Note that `sha_len` can only be in {224, 256, 384, 512} for KMAC HWIP, these
+ * lengths are explicitly passsed in from SV code, so no need to use an
+ * assertion here.
+ */
+void get_sha3_digest(uint64_t sha_len, const svOpenArrayHandle msg,
+                     uint64_t msg_len, svOpenArrayHandle digest) {
+  /**
+   * Number of bytes in result digest
+   */
+  uint64_t digest_len = sha_len / 8;
+
+  uint8_t digest_arr[digest_len];
+
+  /**
+   * Load message from SV memory
+   */
+  uint8_t *msg_arr = (uint8_t *)malloc(msg_len * sizeof(uint8_t));
+  load_arr_from_simulator(msg, msg_arr, msg_len);
+
+  // Compute the digest
+  digestpp::sha3 sha3(sha_len);
+  sha3.absorb(msg_arr, msg_len);
+  sha3.digest(digest_arr, sizeof(digest_arr));
+
+  free(msg_arr);
```

Comment:
malloc() with msg_len == 0 can return NULL, as can other malloc problems. Better always check:

```suggestion
if (msg_arr != nullptr) {
  free(msg_arr);
}
```

Reply:
ahh thats a good point, I can add the checks.